skyportal/handlers/api/invitations.py

```diff
@@ -133,3 +133,136 @@ def post(self):
                 "per their setup docs."
             )
         return self.success()
+
+    @permissions(["Manage users"])
+    def get(self):
+        """
+        ---
+        description: Retrieve invitations
+        parameters:
+          - in: query
+            name: includeUsed
+            schema:
+              type: boolean
+            description: |
+              Bool indicating whether to include used invitations.
+              Defaults to false.
+        responses:
+            200:
+              content:
+                application/json:
+                  schema:
+                    allOf:
+                      - $ref: '#/components/schemas/Success'
+                      - type: object
+                        properties:
+                          data:
+                            type: array
+                            items:
+                              type: array
+                              items:
+                                $ref: '#/components/schemas/Invitation'
+        """
+        include_used = self.get_query_argument("includeUsed", False)
+        query = Invitation.query
+        if not include_used:
+            query = query.filter(Invitation.used.is_(False))
+        invitations = query.all()
+        return_data = [invitation.to_dict() for invitation in invitations]
+        for idx, invite_dict in enumerate(return_data):
+            invite_dict["streams"] = invitations[idx].streams
+            invite_dict["groups"] = invitations[idx].groups
+            invite_dict["invited_by"] = invitations[idx].invited_by
+
+        return self.success(data=return_data)
+
+    @permissions(["Manage users"])
+    def patch(self, invitation_id):
+        """
+        ---
+        description: Update a pending invitation
+        requestBody:
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  groupIDs:
+                    type: array
+                    items:
+                      type: integer
+                  streamIDs:
+                    type: array
+                    items:
+                      type: integer
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+        """
+        data = self.get_json()
+        invitation = Invitation.query.get(invitation_id)
+        group_ids = data.get("groupIDs")
+        stream_ids = data.get("streamIDs")
+        if group_ids is None and stream_ids is None:
+            return self.error(
+                "At least one of either groupIDs or streamIDs are requried."
+            )
+        if group_ids is not None:
+            groups = Group.query.filter(Group.id.in_(group_ids)).all()
+            if len(groups) != len(group_ids):
```

Comment:
Can we do something like a set difference here to return the specific invalid group IDs in the error message? Same thing with the stream IDs right below

Reply:
Sure